README
Net-Int-Stats version 2.07
===========================

NAME
    Net::Int::Stats - Reports specific ifconfig values for a network interface

SYNOPSIS
    use Net::Int::Stats;

    my $get = Net::Int::Stats->new();

    # get value for specific interface
    my $int     = 'eth0';
    my $stat	= 'rx_packets';
    my $packets = $get->value($int, $stat);

DESCRIPTION
    This module provides various statistics generated from the ifconfig command for specific interfaces. 
    RX values consist of packets, errors, dropped, overruns, and frame. TX values consist of packets, 
    errors, dropped, overruns, and carrier. In addition, collisions and txqueuelen are reported. Values 
    are in the format of type:n - ex 'packets:123456'.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module is platform dependent. It uses the linux version
of /sbin/ifconfig. Other platforms such as the windows equivalent
of ipconfig, mac osx, and other versions of unix are not supported. 
This is due to the fact that each platform generates and displays
different information in different formats of ifconfig results.
The linux version is used over the other platforms because of the 
amount of data the default command outputs.

SEE ALSO

linux command /sbin/ifconfig

NOTES

ifconfig output contains more information than the values that are
extracted in this module. More values and/or support for other
operating systems can be added if there are any requests to do so.

COPYRIGHT AND LICENSE

Copyright (C) 2009 Bruce Burch

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.



Changes
Revision history for Perl extension Net::Int::Stats.

1.00  Sat Jun 27 01:51:53 2009
	- original version; created by h2xs 1.23 with options
		-XAn Net::Int::Stats

# versions 1.01 and 1.02 were documentation typo corrections

1.03 Tues Jul 07 12:28:00 2009
	- added operating system check for make test

2.0  Tues Sep 22 11:30:00 2009
	- major code revision
	- still performs the same purpose
	- one method now replaces the previous twelve
	- the user now has to supply two arguments instead of one
	- not quite as user friendly as before with the extra 
	- argument but the code has been tightened up considerably
	- the extracting of values has been refactored

2.01 Thurs Sep 24 13:38:00 2009
	- minor administrative changes
	- updated version in README to 2.0
	- fixed indenting in code
	- small change in pod doc methods section

2.02 Sat Sep 26 15:04:00 2009
	- no functionality changes
	- prior code indentation fix didn't work
	- could have left as is but it's not great for readability
	- used spaces instead of tabs to overcome different tab defaults
	- between vi configs

2.03 Fri Apr 16 16:20:44 PDT 2010
	- no functionality changes
	- fixed typo in synopsis
	- updated version in README

2.04 Sat Jul  2 16:05:38 PDT 2011
	- no functionality changes
	- added more comments
	- refactored part of code that generates the values data
	- placed in it's own subroutine: data()

2.05 Sun Jul  3 18:02:40 PDT 2011
	- no functionality changes
	- fixed pesky space/tab issue formatting in code and comments
	- doesn't affect anything other than readability

2.06 Fri Jul  8 13:57:07 PDT 2011
	- fixed a bug introduced in v2.04
	- the new data() subroutine added to the value() method was
	- being called after the validate() subroutine causing validation
	- against an empty list
	- fixed by calling data() before validate()

2.07 Mon Aug  1 14:19:08 PDT 2011
	- no functionality changes
	- there are FAIL reports for the unsupported operating systems
	- even though install properly bails out during make test for the
	- unsupported operating systems
	- to try to get N/A reports instead of FAIL the bail out text in
	- the test file has been changed to 'OS unsupported' as per
	- suggestion in CPAN Testers Wiki


lib/Net/Int/Stats.pm
package Net::Int::Stats;

our $VERSION = '2.07';

use strict;
use warnings;

############ Global Declarations ##############

# store ifconfig output
my @interface = `/sbin/ifconfig`;

# hash of hashes
# key1 - interface, key2 - type ex: rx_packets, values ex: 'packets:12345'
my %interfaces;

# tmp array to store string tokens
my @tmp;

# interface name
my $key1;

# value types
my @key2;

########## End Global Declarations ###########

# generate ifconfig values
sub data {

    # loop through each line of ifconfig output
    foreach (@interface){

        # skip if blank line
        next if /^$/;

        # get interface name if not white space
        if (!/^\s/){

            # extract values
            extract($_);

            # store first token of interface name
            $key1 = shift(@tmp);
        }

        # get RX, TX, collisions and txqueuelen values
        # look for 'RX' or 'TX' or 'collisions' text
        if (/RX packets/ || /TX packets/ || /collisions/){

            # key2 values
            @key2 = qw(rx_packets rx_errors rx_dropped rx_overruns rx_frame) if (/RX packets/);
            @key2 = qw(tx_packets tx_errors tx_dropped tx_overruns tx_carrier) if (/TX packets/);
            @key2 = qw(collisions txqueuelen) if (/collisions/);

            # extract values
            extract($_);

            # shift first token of 'RX' or 'TX'
            shift(@tmp) if (/RX packets/ || /TX packets/);

            # build values hash
            build();
        }
    }
}

# extract values
sub extract {

    # ifconfig output line
    my $line = shift;

    # remove spaces
    $line =~ s/^\s+//;

    # store tokens split on spaces
    @tmp = split (/\s/, $line);
}

# build values hash
sub build {

    # values type count
    my $i = 0;

    # loop through value types
    for (@key2){
	
        # build hash with interface name, value type, and value
        $interfaces{$key1}{$_} = $tmp[$i];

        # increment values type count
        $i++;
    }
}

# validate interface name
sub validate {

    # interface name
    my $int = shift;

    # terminate program if specified interface name is not in ifconfig output
    die "specified interface $int not listed in ifconfig output!\n" if !(grep(/$int/, keys %interfaces));
}

# create new Net::Int::Stats object
sub new {

    # class name
    my $class = shift;

    # allocate object memory
    my $self  = {};

    # assign object reference to class
    bless($self, $class);

    # initialize values reference
    $self->{VALUES} = '';

    # return object reference
    return $self;
}

# get specific ifconfig value for specific interface
sub value {

    # object reference
    my $self = shift;

    # interface name
    my $int  = shift;

    # value type
    my $type = shift;

    # generate value data
    data();

    # validate if supplied interface is present
    validate($int);

    # user specified value
    $self->{VALUES} = $interfaces{$int}{$type};

    # return value
    return $self->{VALUES};
}

1;

__END__

=head1 NAME

Net::Int::Stats - Reports specific ifconfig values for a network interface

=head1 SYNOPSIS

  use Net::Int::Stats;

  my $get = Net::Int::Stats->new();

  # get value for specific interface
  my $int     = 'eth0';
  my $stat    = 'rx_packets';
  my $packets = $get->value($int, $stat);

=head1 DESCRIPTION

This module provides various statistics generated from the ifconfig command for specific interfaces. 
RX values consist of packets, errors, dropped, overruns, and frame. TX values consist of packets, 
errors, dropped, overruns, and carrier. In addition, collisions and txqueuelen are reported. Values 
are in the format of type:n - ex 'packets:123456'.

=head1 METHODS

Use this one method to get specific values which requires two arguments: B<value()>.
Ex: $packets = $get->value($int, 'rx_packets');

The first argument is the interface and the second is the type value to extract.

RX values - rx_packets, rx_errors, rx_dropped, rx_overruns, rx_frame

TX values - tx_packets, tx_errors, tx_dropped, tx_overruns, tx_carrier

Miscellaneous values - collisions, txqueuelen

=head1 DEPENDENCIES

This module is platform dependent. It uses the linux version
of /sbin/ifconfig. Other platforms such as the windows equivalent
of ipconfig, mac osx, and other versions of unix are not supported. 
This is due to the fact that each platform generates and displays
different information in different formats of ifconfig results.
The linux version is used over the other platforms because of the 
amount of data the default command outputs.  

=head1 SEE ALSO

linux command /sbin/ifconfig

=head1 NOTES

ifconfig output contains more information than the values that are
extracted in this module. More values and/or support for other
operating systems can be added if there are any requests to do so.   

=head1 AUTHOR

Bruce Burch <bcb12001@yahoo.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Bruce Burch

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut
